### PR TITLE
Issue2780 - getValue() for special telemetry fields

### PR DESCRIPTION
**Pull request by dsbeach**
Changes to address issue #2780.  Tested on Taranis hardware using two different GPS units and 2-cell and 3-cell lipos with FrSky Lipo voltage sensor.

Note: this change also includes modification of getValue for UNIT_GPS to use table members 'lat' and 'lon' (not capitalized) following the convention of the luaGetDateTime() function.

Due to problems related to variable scope, I opted to comment the initial call to getValue(src) indicating that it is ignored for special telemetry values rather than restructure the use of the variable 'value'.

David

### CODE REVIEW

File: radio/src/lua_api.cpp [hunk 1]
```diff
@@ -144,31 +150,57 @@ static void luaPushLatLon(TelemetrySensor & telemetrySensor, TelemetryItem & tel
...
+  luaPushDateTime(L, telemetryItem.datetime.year + 2000, telemetryItem.datetime.month, telemetryItem.datetime.day,
+                  telemetryItem.datetime.hour, telemetryItem.datetime.min, telemetryItem.datetime.sec);
+}
+
+static void luaPushCells(TelemetrySensor & telemetrySensor, TelemetryItem & telemetryItem)
+{
+  if (telemetryItem.cells.count == 0)
+    lua_pushinteger(L, (int)0); // returns zero if no cells
+  else {
+    lua_createtable(L, 0 /*telemetryItem.cells.count*/, 0);
```
Comment (projectkk2glider) [inline radio/src/lua_api.cpp]: Why did you comment out the `telemetryItem.cells.count`, it should help with memory allocation.

**Comment by projectkk2glider** [comment]
Looks good. I have not tested this.

File: radio/src/lua_api.cpp (hunk 1 above)
Comment (dsbeach) [inline radio/src/lua_api.cpp]: The change was made after my initial testing when I noticed that the convention for this module seemed to me using the lua_newtable() macro (which expands to lua_createtable(L, 0, 0)).

There are 13 references to lua_newtable() in this file.  Should they all be changed to explicit allocations?

David

Comment (projectkk2glider) [inline radio/src/lua_api.cpp]: >There are 13 references to lua_newtable() in this file. Should they all be changed to explicit allocations?

Well probably no, but when you explicitly call lua_createtable()  you can optimize the call by providing end table size. Its not a big issue in any case.

**Comment by dsbeach** [comment]
Now using pre-allocated table space via lua_createtable() for GPS, DATETIME, and CELLS.

**Comment by bsongis** [comment]
Thanks!

**Comment by galosre** [comment]
Maybe not relevant to this thread but for 4 and 6 cells I think there is an improvement but it not completely fixed as for me the value now between 4.01 and 4.09 still do not play correctly it omitted the point and the middle digit I use this script modified by a Swiss guy (Sacre100) but before it use to do that even for all value above 4.00v up to 4.19 anyway below 3.99 everything is correct

    --[[Quick and dirty two decimal place voice
    announcement for Cels or other telemetry cell
    voltages that are reported at 0.01V resolution

    Put the script in SD/SCRIPTS/FUNCTIONS folder.
    Specify the input channel name and logical or
    physical switch that triggers script below.
    Then set up a SF to play this script when  
    mySwitch is activated.]]

    myVoltage="Cels"

    local delay_val = 500 --minimum time between announcements (x10ms)
    local wake_time = 0

    local function sum_cels(cels)
      local val = 0
      for i, v in pairs(cels) do
        val = val + v
      end
      return val
    end

    local function announce(voltage)
      local cell = getValue(voltage)
      if "table" == type(cell) then
        cell = sum_cels(cell)
      end
      if cell and cell ~= 0 then
        playNumber(math.floor(cell * 10), 0 ,PREC1)
        if math.floor(cell * 10 % 10) == 0 then
          playFile("SYSTEM/0180.wav")
        end
        playNumber(cell * 100 % 10, 1)
      else
        print("Le capteur " .. voltage .. " n'existe pas")
      end
      wake_time = getTime() + delay_val
    end

    local function run(event)
      if getTime() >= wake_time then
        announce(myVoltage)
      end
    end

    return {run = run}



whole discussion below

 https://translate.google.com/translate?sl=fr&tl=en&js=y&prev=_t&hl=en&ie=UTF-8&u=http%3A%2F%2Ffrskytaranis.forumactif.org%2Ft3984-nouveau-ici-du-bout-du-monde&edit-text=

**Comment by 3djc** [comment]
You need to discuss that with the script author./maintainer

**Comment by kilrah** [comment]
That script hardcodes "SYSTEM/0180.wav" as being "point zero". That's correct for the French voice pack on 2.1 for which this "quick and dirty" script as labelled was intended for, but it doesn't handle the fact that other languages and other versions use different numbering.
So modify the script to your setup.

**Comment by galosre** [comment]
I knew it was not really relevant but thousand times thank you @ kilrah  for putting on the right way in order for me to be able to correct this error.
je ne le ferai plus!

**Pull request merged**